eve/validation.py

```diff
@@ -12,5 +12,144 @@
+import cerberus
+import cerberus.errors
+from cerberus import DocumentError, SchemaError  # flake8: noqa
+
+from eve.utils import config
+
+
+class Validator(cerberus.Validator):
+
+    def __init__(self, *args, **kwargs):
+        kwargs['error_handler'] = DefaultErrorHandler
+
+        resource = kwargs.get('resource', None)
+        if resource:
+            resource_def = config.DOMAIN[resource]
+            kwargs['allow_unknown'] = resource_def['allow_unknown']
+        super(Validator, self).__init__(*args, **kwargs)
+
+    def validate_update(self, document, document_id, persisted_document=None):
+        """ Validate method to be invoked when performing an update, not an
+        insert.
+
+        :param document: the document to be validated.
+        :param document_id: the unique id of the document.
+        :param persisted_document: the persisted document to be updated.
+        """
+        self.document_id = document_id
+        self.persisted_document = persisted_document
+        return super(Validator, self).validate(document, update=True)
+
+    def validate_replace(self, document, document_id, persisted_document=None):
+        """ Validation method to be invoked when performing a document
+        replacement. This differs from :func:`validation_update` since in this
+        case we want to perform a full :func:`validate` (the new document is to
+        be considered a new insertion and required fields needs validation).
+        However, like with validate_update, we also want the current document_id
+        not to be checked when validating 'unique' values.
+
+        :param document: the document to be validated.
+        :param document_id: the unique id of the document.
+        :param persisted_document: the persisted document to be updated.
+
+        .. versionadded:: 0.1.0
+        """
+        self.document_id = document_id
+        self.persisted_document = persisted_document
+        return super(Validator, self).validate(document)
+
+    def _normalize_default(self, mapping, schema, field):
+        """ {'nullable': True} """
+        if not self.persisted_document or \
+           field not in self.persisted_document:
+            super(Validator, self)._normalize_default(mapping, schema, field)
+
+    def _normalize_default_setter(self, mapping, schema, field):
+        """ {'oneof': [
+                {'type': 'callable'},
+                {'type': 'string'}
+                ]} """
+        if not self.persisted_document or \
+           field not in self.persisted_document:
+            super(Validator, self)._normalize_default_setter(mapping, schema,
+                                                             field)
+
+    def _validate_dependencies(self, dependencies, field, value):
+        """ {'type': ['dict', 'hashable', 'hashables']} """
+        persisted = self._filter_persisted_fields_not_in_document(dependencies)
+        if persisted:
+            dcopy = copy.copy(self.document)
```

Comment:
I'm leaving this for later. I'm not familiar enough with the code to be able to make an informed change here.